OpenSim/Common/Component.h

```diff
@@ -925,24 +954,27 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
     void dumpSubcomponents(int depth=0) const;
 
 protected:
-
-    /** Construct Component as a subcomponent member of another Component.
-    Automatically marks this component as a subcomponent of the owner.
-    A subcomponent MUST be named. **/
-    explicit Component(const std::string& name, Component* owner);
-
+    class StateVariable;
+    //template <class T> friend class ComponentSet;
+    // Give the ComponentMeasure access to the realize() methods.
+    template <class T> friend class ComponentMeasure;
+
+#ifndef SWIG
+    /// @class MemberSubcomponentIndex
+    /// Unique integer type for local member subcomponent indexing
+    SimTK_DEFINE_UNIQUE_INDEX_TYPE(MemberSubcomponentIndex);
+
+    /** Construct a subcomponent as a data member of this Component. All Component
+        interface calls are automatically invoked on its subcomponents. */
     template<class C=Component>
-    C& constructSubcomponent(const std::string& name) {
+    MemberSubcomponentIndex constructSubcomponent(const std::string& name) {
         C* component = new C();
         component->setName(name);
-        this->markAsSubcomponent(component);
-        return *component;
+        component->setParent(*this);
+        _memberSubcomponents.push_back(SimTK::ClonePtr<Component>(component));
+        return MemberSubcomponentIndex(_memberSubcomponents.size()-1);
```

Comment:
I understand now why the member can't be a reference.

If you add those accessors by index, then Tom can use them in his PR.